CONTRIBUTING.md

```diff
@@ -1,12 +1,379 @@
+```
+
+receives the event and handles updating the attribute on the correct zigbee cluster. As you can see there really isn't much here that needs to be done to accomplish our goal.
+
+Once we have created our `CustomCluster` implementations we have to tell the `CustomDevice` implementation to use them. We do this in the `replacement` dict in the quirk definition. Start by copying the `signature` dict and remove the `models_info` from it. Then we replace the cluster ids that we want to override with the names of our `CustomCluster` implementations that we have created. The result looks like this:
+
+```python
+replacement = {
+    SKIP_CONFIGURATION: True,
+    ENDPOINTS: {
+        1: {
+            PROFILE_ID: zha.PROFILE_ID,
+            DEVICE_TYPE: zha.DeviceType.SMART_PLUG,
+            INPUT_CLUSTERS: [
+                BasicCluster,
+                PowerConfiguration.cluster_id,
+                DeviceTemperature.cluster_id,
+                Groups.cluster_id,
+                Identify.cluster_id,
+                OnOff.cluster_id,
+                Scenes.cluster_id,
+                BinaryOutput.cluster_id,
+                Time.cluster_id,
+                ElectricalMeasurementCluster,
+            ],
+            OUTPUT_CLUSTERS: [Ota.cluster_id, Time.cluster_id],
+        },
+        2: {
+            PROFILE_ID: zha.PROFILE_ID,
+            DEVICE_TYPE: zha.DeviceType.MAIN_POWER_OUTLET,
+            INPUT_CLUSTERS: [AnalogInputCluster],
+            OUTPUT_CLUSTERS: [AnalogInput.cluster_id, Groups.cluster_id],
+        },
+        3: {
+            PROFILE_ID: zha.PROFILE_ID,
+            DEVICE_TYPE: zha.DeviceType.METER_INTERFACE,
+            INPUT_CLUSTERS: [AnalogInput.cluster_id],
+            OUTPUT_CLUSTERS: [AnalogInput.cluster_id],
+        },
+        100: {
+            PROFILE_ID: zha.PROFILE_ID,
+            DEVICE_TYPE: zha.DeviceType.OCCUPANCY_SENSOR,
+            INPUT_CLUSTERS: [BinaryInput.cluster_id],
+            OUTPUT_CLUSTERS: [BinaryInput.cluster_id, Groups.cluster_id],
+        },
+```
+
+You can see that we have replaced `ElectricalMeasurement.cluster_id` from endpoint 1 in the `signature` dict with the name of our cluster that we created: `ElectricalMeasurementCluster` and on endpoint 2 we replaced `AnalogInput.cluster_id` with the implementation we created for that: `AnalogInputCluster`. This instructs Zigpy to use these `CustomCluster` derivatives instead of the normal cluster definitions for these clusters and this is why this part of the quirk is called `replacement`.
+
+Now lets put this all together. If you examine the device definition above you will see that we have defined our custom device, we defined the `signature` dict where we transcribed the `SimpleDescriptor` output we obtained when the device joined the network and we defined the `replacement` dict where we swapped the cluster ids for the culsters that we wanted to replace with the `CustomCluster` implementations that we created.
+
```

Comment:
While working on my (non-working, ongoing) attempt on the orvibo motion detector, I noticed that Xiaomi had added in the `replacement` either an extra (occupancy or motion) cluster - this kind of makes sense as one is derivative of the other and adding the cluster makes the component more viable.

Does this imply the actual `signature` supported value must publish to the unsupported `replacement` value?

Reply:
I'm not quite following what are you asking. Perhaps references to specific examples/code would help.

Consider `signature` as a must have match, for the `replacement` to kick in. The content of replacement doesn't have to match the signature, it is entirely up to quirk author and the way device is functioning, but in 99% of the use cases, the replacement is a copy of signature with some attributes/clusters replaced by the quirk.
e.g https://github.com/zigpy/zha-device-handlers/blob/e09f7f81171de15c84825f5e0b3f3b987616f829/zhaquirks/sinope/light.py#L67 just changes the device type, so ZHA discovers that device a "light" and not as a "switch", leaving everything else as is. 

In other cases, replacement may contain extra "virtual" cluster not present on the original devices, like ElectricalMeasurement cluster for Xiaomi devices. Xiaomi reports power usage on a analog cluster (IIRC), but ZHA doesn't know about power metering on analog cluster. So quirks gets data from analog cluster and presents it through a virtual "ElectricalMeasurement" cluster which ZHA knows how to represent.